cpp/src/barretenberg/dsl/acir_format/recursion_constraint.cpp

```diff
@@ -0,0 +1,346 @@
+                             field_ct::from_witness_index(&composer, aggregation_input[4 * i + 1]),
+                             field_ct::from_witness_index(&composer, aggregation_input[4 * i + 2]),
+                             field_ct::from_witness_index(&composer, aggregation_input[4 * i + 3]));
+            aggregation_elements[i].assert_is_in_field();
+        }
+        // If we have a previous aggregation object, assign it to `previous_aggregation` so that it is included
+        // in stdlib::recursion::verify_proof
+        previous_aggregation.P0 = bn254::g1_ct(aggregation_elements[0], aggregation_elements[1]);
+        previous_aggregation.P1 = bn254::g1_ct(aggregation_elements[2], aggregation_elements[3]);
+        previous_aggregation.has_data = true;
+    } else {
+        previous_aggregation.has_data = false;
+    }
+
+    transcript::Manifest manifest = Composer::create_unrolled_manifest(input.public_inputs.size());
+
+    std::vector<field_ct> key_fields;
+    key_fields.reserve(input.key.size());
+    for (const auto& idx : input.key) {
+        key_fields.emplace_back(field_ct::from_witness_index(&composer, idx));
+    }
+
+    std::vector<field_ct> proof_fields;
+    proof_fields.reserve(input.proof.size());
+    for (const auto& idx : input.proof) {
+        proof_fields.emplace_back(field_ct::from_witness_index(&composer, idx));
+    }
+
+    // recursively verify the proof
+    std::shared_ptr<verification_key_ct> vkey = verification_key_ct::from_field_elements(
+        &composer, key_fields, inner_proof_contains_recursive_proof, nested_aggregation_indices);
+    vkey->program_width = noir_recursive_settings::program_width;
+    Transcript_ct transcript(&composer, manifest, proof_fields, input.public_inputs.size());
+    aggregation_state_ct result = proof_system::plonk::stdlib::recursion::verify_proof_<bn254, noir_recursive_settings>(
+        &composer, vkey, transcript, previous_aggregation);
+
+    // Assign correct witness value to the verification key hash
+    vkey->compress().assert_equal(field_ct::from_witness_index(&composer, input.key_hash));
+
+    ASSERT(result.public_inputs.size() == input.public_inputs.size());
+
+    // Assign the `public_input` field to the public input of the inner proof
+    for (size_t i = 0; i < input.public_inputs.size(); ++i) {
+        result.public_inputs[i].assert_equal(field_ct::from_witness_index(&composer, input.public_inputs[i]));
+    }
+
+    // Assign the recursive proof outputs to `output_aggregation_object`
+    for (size_t i = 0; i < result.proof_witness_indices.size(); ++i) {
+        const auto lhs = field_ct::from_witness_index(&composer, result.proof_witness_indices[i]);
+        const auto rhs = field_ct::from_witness_index(&composer, input.output_aggregation_object[i]);
+        lhs.assert_equal(rhs);
+    }
+}
+
+template void create_recursion_constraints<false>(Composer&, const RecursionConstraint&);
```

Comment:
This differs from how it is done for ecdsa_constraints -- we should add dummy_constraints in one way

Reply:
I'll switch from template to a flag